### ISSUE TITLE
Compatibility with Anaconda

### ISSUE THREAD

**Issue by moonshoes87** [PmagPy/PmagPy#84]
As we start contemplating the switch to Python 3, maybe we can revisit the two issues with Anaconda.  It is unclear when Canopy will support Python 3, so I'd like to have other options.  

To recap:

Problem 1:
I’m also getting some segmentation faults with the GUIs.  This seems to happen consistently with Pmag GUI step 3 (adding er_\* information).  

/anaconda/bin/pythonw: line 3:  2366 Segmentation fault: 11  //anaconda/python.app/Contents/MacOS/python "$@"

Problem 2:
nebula$ python pmag_gui.py
-I- Starting Pmag GUI - please be patient
This program needs access to the screen.
Please run with a Framework build of python, and only when you are
logged in on the main display of your Mac.

@Swanson-Hysell  or @Caoimhinmg, if one of you could check and see if problem 1 is still occurring (I no longer have Anaconda installed), that would be awesome!  Regarding problem 2, I can't remember if the framework build issue affected the pip install.  With Anaconda, have you been able to do the pip install and run the command line programs?

**Comment by Caoimhinmg** [comment]
Well I don't have a mac so I can't check Anaconda on OSX I'll leave that to @Swanson-Hysell , but I dual boot an Ubuntu based Linux os (Linux Mint 17.3) and Windows 7. I tried step 3 of Pmag GUI on 2 different data sets in both OSs and it worked fine. Just for completeness my Windows 7 distro is running Anaconda based python2 and my Linux is running python2 and python3 side along using apt the Ubuntu standard package manager with pip used to upgrade some of the packages as the Linux default ones are sometimes older for stability reasons.

In short no segfaults over here. As to the access to the screen thing I've got no idea the whole python vs. pythonw interpreter seems to be a mac thing I've never understood. I'll see what I can dig up on it though and try the command line programs installed via pip next time I boot into my Windows 7 distro.

**Comment by Swanson-Hysell** [comment]
With regards to motivating this Python 3 switch this page is useful: https://python3statement.github.io The entire scientific python stack will drop support for Python 2.7 before long. Pandas just joined this statement too even though they aren't listed yet.

Anaconda really does seem to be becoming the package manager of choice among python using academics. 

@yuempark ran into the framework build issue and pushed a related change to the Cookbook with this pythonw workaround: 

```
% pythonw ./pmag_gui.py
```

I do get the segmentation fault issue using Anaconda, which is a mystery to me. Here is the error:

```
/Users/Laurentia/anaconda/envs/python2/bin/pythonw: line 3: 20795 Segmentation fault: 11  /Users/Laurentia/anaconda/envs/python2/python.app/Contents/MacOS/python "$@"
```

One question is whether this segmentation fault will be an error with everything in Python 3 (for which the wxPython Pheonix implementation would need to be used: https://github.com/wxWidgets/Phoenix). Kevin, you want to revamp your Python 3 branch and I can run Pmag_GUI on my Python 3 anaconda installation (after installing wxPhoenix) and I can see if I get the issue.

Also, @moonshoes87 I think it could be helpful for these purposes if you do get a Anaconda Python3 distro on you local machine. I have both Canopy and Anaconda going and switch between them by commenting/uncommenting lines in my .bash_profile

**Comment by Caoimhinmg** [comment]
K, @Swanson-Hysell just rolled back and forcibly reconverted my fork of PmagPy same as I did the first time I translated the whole repo to Python3. I have tested all the GUI's on my Linux distribution and it seems to be working right now. Though the Pmag_GUI step three seems to having some trouble getting the magic data model and has made running all the tests nearly impossible (ran them all weekend and they didn't terminate). Demag_GUI is passing all its tests and Pmag_GUI and Thellier_GUI work when I run them, I think the test issue is a directory problem which is of course still being resolved here (#77).

Anyway I rolled back to d5a387c114886b298fe1ba4205261057ccea6043 before converting so I'm a bit behind master because it took some time to convert. I'm not merging with upstream/master because the python2 code is what destroyed the last conversion I did. 

As for the installing WxPython Phoenix I can't be too much help on mac which seems to be the one OS you have to build from source on (there are unofficial pip installs for linux and windows). Here's a small guide of what I've found:

* Mac - you basically have to build from the github repo, here https://github.com/wxWidgets/Phoenix#how-to-build-phoenix, so you'll have the most up to date stuff but it's also the hardest install. There are also a few dependencies that don't come packaged so you'll have to read the readme install guide in detail to make sure you have all of them.
* Windows - There is a pip install script here: https://wiki.wxpython.org/How%20to%20install%20wxPython
* Linux (Ubuntu) - If there is anyone else using an Ubuntu based distro here's my install script just paste into a terminal: 

```
sudo apt-get install dpkg-dev build-essential python2.7-dev python3-dev libwebkitgtk-dev libjpeg-dev libtiff-dev libgtk2.0-dev libsdl1.2-dev libgstreamer-plugins-base0.10-dev libnotify-dev freeglut3 freeglut3-dev python3 python3-pip
sudo pip3 install --upgrade pip
sudo -H pip3 install --upgrade setuptools
sudo -H pip3 install --upgrade --trusted-host wxpython.org --pre -f http://wxpython.org/Phoenix/snapshot-builds/ wxPython_Phoenix
```

Now if you're having trouble running the programs and want to fix them here are the guides I've been using for translating classic WxPython to Phoenix:

* https://wxpython.org/Phoenix/docs/html/MigrationGuide.html
* https://wxpython.org/Phoenix/docs/html/classic_vs_phoenix.html

So now @Swanson-Hysell you should be able to test for that segfault in Python3 a lot of things the lower lying machine interface is different in Python3 so hopefully this fixes the issue.

**Comment by njarboe** [comment]
I have just got a new Mac laptop and will be installing Python and PmagPy shortly. I would like to follow the directions that the general public uses (cookbook) to install Python and PmagPy, but if we are going to move to Anaconda, this might be a good opportunity to test for that version of Python on a clean machine. I could write instructions on the Anaconda install procedure and change the cookbook, if that is what we want people to use going forward.

**Comment by moonshoes87** [comment]
I’ve been playing with Anaconda on my machine, and here’s the current status:  

I’m trying to track down a seg fault in Pmag GUI with Anaconda.  Because of this breaking error, I don’t think we can recommend Anaconda officially yet.  But I’m working on it.  Anaconda is probably going to become more important when we switch to Python 3, which is still down the road.

If you do want to install Anaconda for PmagPy, these are the steps that I found to be necessary:
# follow regular Anaconda install directions from https://docs.continuum.io/anaconda/install
conda install wxpython
conda install -c omnia scripttest=1.3

I haven’t worked with the pip installation of pmagpy/pmagpy-cli and Anaconda yet, as I am still working on the one error I mentioned above.  I just have these lines in my .bashrc file:	

PATH=/Users/***/Python/PmagPy:$PATH
PATH=/Users/***/Python/PmagPy/programs:$PATH
export PYTHONPATH=$PYTHONPATH:/Users/***/Python/PmagPy/

Then, if you want to run the tests from the PmagPy directory, you need to use:

pythonw -m unittest discover   # pythonw allows anaconda to access the screen

I plan to test the pip installations with Anaconda ASAP.

Nick, if I were you I think I would stick with the Canopy/pip installation for now and see if you have any problems getting that to work for you, since we aren’t yet ready to switch over to Anaconda.  It would be good to make sure our install directions are 1) relatively clear and 2) working.

**Comment by moonshoes87** [comment]
Ok, seg fault is fixed.  A window was being destroyed while it was still used in a sizer -- somehow the events must shake out differently with Canopy, I don't know why.  Now all of the basic PmagPy code seems to work correctly with Anaconda, so that's a good step!

However, using pip install with Anaconda breaks the GUIs, with this error:

This program needs access to the screen.
Please run with a Framework build of python, and only when you are
logged in on the main display of your Mac.


For some reason, the environment is correctly set to use pythonw if you manually add PmagPy and PmagPy/programs to your PATH (because of this line at the top of each GUI file:  #!/usr/bin/env pythonw  ), but that doesn't seem to translate with using setuptools and installing with pip.  

I'm still working on this, if anyone has suggestions let me know.

**Comment by moonshoes87** [comment]
Ok, I found a hack-y solution.  

Instead of using setuptools "entry_points" (the preferred approach for making executables) I can use setuptools "scripts", which then allows me to write a custom setup script to launch one of the GUIs using pythonw.  The downside with this approach is that I use this line to set the environment:

/usr/bin/env pythonw

which grabs whatever Python is in the $PATH.  This is good in development but bad in production.  Pip generally installs python scripts with the path to the Python you used to download that package hard-coded at the top, so:  /Users/nebula/anaconda2/bin/python  instead of /usr/bin/env python.  What this means is that our installed script might be more prone to breaking if the user tries a different Python that isn't compatible.  explained better by someone else here:

#!/usr/bin/env python executes the system's default Python version. That is fine as long as your code is compatible with said version. When the default version is updated (from 2.5 to 3, say) your code or other Python code which references /usr/bin/env may stop working, even though the old Python version is still installed. For that reason it makes sense to "hardcode" the path to the appropriate python interpreter.    http://stackoverflow.com/questions/1530702/dont-touch-my-shebang

Anyway, what this means to me is that we should add Anaconda compatible scripts (that may be slightly less stable) to the existing pip package.  From a UI perspective, what should we call those scripts?  In testing I called the executables things like pmag_gui_a (a for anaconda), but maybe something else would be better?  (Note that we can't use a command line argument here to specify anaconda vs non-anaconda -- it has to be the actual name of the executable).  I would love input on this naming issue.  I have pushed to master changes in command_line_setup.py and adding a bin/ directory with the Anaconda scripts if anyone wants to have a look.  

You can also pip install a mostly-Anaconda compatible pmagpy-cli from the test PyPI site:  

 pip install -i https://testpypi.python.org/pypi pmagpy-cli --upgrade 

That install should work with Anaconda with the commands pmag_gui_a, demag_gui_a, etc.

Also, I've found several more incompatibilities with the Anaconda implementations of the GUIs.  I'm going to keep working on those problems, and create new issues for any bugs that I need help with.  Slow progress!

**Comment by njarboe** [comment]
I would say for clarity you could call it pmag_gui_anaconda, etc. until
anaconda is the default for python installation and then revert to
pmag_gui. Maybe a bit long, but it would be clear what one was invoking.

**Comment by moonshoes87** [comment]
I've added pmag_gui_anaconda (etc.) to pmagpy-cli and to the Cookbook documentation.  It is still the hack-y solution as above, but it seems to work.  I don't know of any other Anaconda-specific issues, so I think we currently fully supporting both Canopy and Anaconda : )
